src/v/cluster/rm_stm.cc

```diff
@@ -2014,50 +2014,56 @@ rm_stm::do_aborted_transactions(model::offset from, model::offset to) {
     co_return result;
 }
 
-void rm_stm::compact_snapshot() {
-    auto cutoff_timestamp = model::timestamp::now().value()
-                            - _transactional_id_expiration.count();
-    if (cutoff_timestamp <= _oldest_session.value()) {
-        return;
+bool rm_stm::is_producer_expired(
+  model::timestamp now, const seq_entry& entry) const {
+    /**
+     * Producer is expired if it has no ongoing transactions and its last write
+     * timestamp is older than requested producer id
+     */
+    auto lock_it = _tx_locks.find(entry.pid.get_id());
+    /**
+     * If there are ongoing transactions we must keep the producer state alive
+     */
+    if (lock_it != _tx_locks.end() && !lock_it->second->ready()) {
+        return false;
     }
 
-    if (_log_state.seq_table.size() <= _seq_table_min_size) {
-        return;
+    if (_log_state.current_txes.contains(entry.pid)) {
+        return false;
     }
 
-    if (_log_state.seq_table.size() == 0) {
+    return now.value() - entry.last_write_timestamp
+           > _transactional_id_expiration.count();
+}
+
+void rm_stm::compact_snapshot() {
+    if (_log_state.seq_table.empty()) {
```

Comment:
nit: may be better to move this check after L2046.. most times seq_table is non empty but we still pass the check in L2045.